File: apps/gateway/src/middlewares/healthcheck.ts
// Avoid depending on external @types/express in this file to prevent
// "Cannot find module 'express' or its corresponding type declarations"
// errors in environments where those types are not installed.
// Define minimal local types sufficient for this middleware.
type Request = {
  path: string;
  // allow indexing for any other properties used elsewhere
  [key: string]: any;
};

type Response = {
  locals: { traceId?: string } & { [key: string]: any };
  status: (code: number) => Response;
  json: (body: any) => Response;
  [key: string]: any;
};

type NextFunction = () => void;

import { createBaseResponse } from "../shared/contracts.js";

export const healthcheckMiddleware = (request: Request, response: Response, next: NextFunction) => {
  if (request.path !== "/health/live") {
    next();
    return;
  }

  response.status(200).json(
    createBaseResponse({
      data: {
        service: "leitesol-gateway",
        status: "healthy",
        version: "v1",
      },
      message: "Gateway liveness check completed.",
      statusCode: 200,
      traceId: response.locals.traceId ?? "gateway-health",
    }),
  );
};

File: apps/api/src/leitesol_api/interfaces/http/middlewares/healthcheck.py
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import JSONResponse, Response

from leitesol_api.infrastructure.settings import get_settings
from leitesol_api.interfaces.responses import build_response_payload


class HealthcheckMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next) -> Response:
        if request.url.path == "/health/live":
            settings = get_settings()
            payload = build_response_payload(
                data={
                    "service": settings.service_name,
                    "status": "healthy",
                    "version": settings.api_version,
                },
                message="API liveness check completed.",
                status_code=200,
                trace_id=request.headers.get("x-request-id"),
            )
            return JSONResponse(status_code=200, content=payload)

        return await call_next(request)
